integration_tests/src/main/python/parquet_test.py

```diff
@@ -779,6 +780,38 @@ def test_spark_32639(std_input_path):
         lambda spark: spark.read.schema(schema_str).parquet(data_path),
         conf=original_parquet_file_reader_conf)
 
+@pytest.mark.skipif(not is_before_spark_320(), reason='Spark 3.1.x does not need special handling')
+def test_spark_40819_31x(std_input_path):
+    data_path = "%s/timestamp-nanos.parquet" % (std_input_path)
+    def read_timestamp_nano_parquet(spark):
+        spark.read.parquet(data_path).collect()
+    # we correctly return timestamp_micros when running against Spark 3.1.x
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: spark.read.parquet(data_path))
+
+@pytest.mark.skipif(is_before_spark_320(), reason='Spark 3.1.x supports reading timestamps in nanos')
+def test_spark_40819_false(std_input_path):
+    data_path = "%s/timestamp-nanos.parquet" % (std_input_path)
+    conf = copy_and_update(original_parquet_file_reader_conf, {
+            'spark.sql.legacy.parquet.nanosAsLong': False })
+    def read_timestamp_nano_parquet(spark):
+        spark.read.parquet(data_path).collect()
+    assert_py4j_exception(
+        lambda: with_gpu_session(read_timestamp_nano_parquet, conf),
+        error_message="Illegal Parquet type: INT64 (TIMESTAMP(NANOS,true))")
```

Comment:
This should verify the Spark CPU also throws an error to verify we're matching CPU behavior, ideally using assert_gpu_and_cpu_error.